### PR TITLE
Better accessibility for reveal

### PR DESCRIPTION
**Pull request by Owlbertz**
I enhanced the accessibility of reveal based on [this issue #6572](https://github.com/zurb/foundation/issues/6579).

### CODE REVIEW

File: doc/includes/reveal/examples_reveal_ajax.html [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {{#markdown}}
 ```html
-<button data-reveal-id="myModal" data-reveal-ajax="http://some-url">
+<button data-reveal-id="myModal" data-reveal-ajax="http://some-url" aria-haspopup="true">
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_ajax.html]: `aria-haspopup` is only used to designate menu-type popups, not required here.

File: doc/includes/reveal/examples_reveal_ajax.html [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
     Click Me For A Modal
 </button>
 
-<a href="http://some-url" data-reveal-id="myModal" data-reveal-ajax="true">
+<a href="http://some-url" data-reveal-id="myModal" data-reveal-ajax="true" aria-haspopup="true" role="button">
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_ajax.html]: This should be a `<button>` tag. `aria-haspopup` and `role="button"` not needed.

File: doc/includes/reveal/examples_reveal_basic.html [hunk 3]
```diff
@@ -6,7 +6,7 @@
   <h2 id="modalTitle">Awesome. I have it.</h2>
   <p class="lead">Your couch.  It is mine.</p>
   <p>I'm a cool paragraph that lives inside of an even cooler modal. Wins!</p>
-  <a class="close-reveal-modal" aria-label="Close">&#215;</a>
+  <a class="close-reveal-modal" aria-label="Close" role="button" tabindex="0">&#215;</a>
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_basic.html]: Making this a `<button>` would be better than making an anchor into a button.

File: doc/includes/reveal/examples_reveal_basic_rendered.html [hunk 4]
```diff
@@ -4,5 +4,5 @@
   <h2 id="modalTitle">Awesome. I have it.</h2>
   <p class="lead">Your couch.  It is mine.</p>
   <p>I'm a cool paragraph that lives inside of an even cooler modal. Wins!</p>
-  <a class="close-reveal-modal" aria-label="Close">&#215;</a>
+  <a class="close-reveal-modal" aria-label="Close" role="button" tabindex="0">&#215;</a>
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_basic_rendered.html]: Making this a <button> would be better than making an anchor into a button.

File: doc/includes/reveal/examples_reveal_data.html [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {{#markdown}}
 ```html
-<a href="#" data-reveal-id="myModal">Click Me For A Modal</a>
+<a href="#" data-reveal-id="myModal" aria-haspopup="true" role="button">Click Me For A Modal</a>
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_data.html]: This should be a <button> tag. aria-haspopup and role="button" not needed.

File: doc/includes/reveal/examples_reveal_modal_examples.html [hunk 6]
```diff
@@ -10,13 +10,13 @@ <h2 id="firstModalTitle">This is a modal.</h2>
   <p>Reveal makes these very easy to summon and dismiss. The close button is simply an anchor with a unicode character icon and a class of <code>close-reveal-modal</code>. Clicking anywhere outside the modal will also dismiss it.</p>
   <p>Finally, if your modal summons another Reveal modal, the plugin will handle that for you gracefully.</p>
   <p><a href="#" data-reveal-id="secondModal" class="secondary button">Second Modal...</a></p>
-  <a class="close-reveal-modal" aria-label="Close">&#215;</a>
+  <a class="close-reveal-modal" aria-label="Close" role="button" tabindex="0">&#215;</a>
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_modal_examples.html]: This should be a <button> tag. aria-haspopup and role="button" not needed. `tabindex="0"` not needed because anchor tags are already in the tab order.

File: doc/includes/reveal/examples_reveal_modal_examples.html [hunk 7]
```diff
@@ -10,13 +10,13 @@ <h2 id="firstModalTitle">This is a modal.</h2>
...
 </div>
 
 <div id="secondModal" class="reveal-modal" data-reveal aria-labelledby="secondModalTitle" aria-hidden="true" role="dialog">
   <h2 id="secondModalTitle">This is a second modal.</h2>
   <p>See? It just slides into place after the other first modal. Very handy when you need subsequent dialogs, or when a modal option impacts or requires another decision.</p>
-  <a class="close-reveal-modal" aria-label="Close">&#215;</a>
+  <a class="close-reveal-modal" aria-label="Close" role="button" tabindex="0">&#215;</a>
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_modal_examples.html]: This should be a tag. aria-haspopup and role="button" not needed. tabindex="0" not needed because anchor tags are already in the tab order.

File: doc/includes/reveal/examples_reveal_modal_examples.html [hunk 8]
```diff
@@ -25,7 +25,7 @@ <h2 id="videoModalTitle">This modal has video</h2>
     <iframe width="1280" height="720" src="//www.youtube-nocookie.com/embed/wnXCopXXblE?rel=0" frameborder="0" allowfullscreen></iframe>
   </div>
 
-  <a class="close-reveal-modal" aria-label="Close">&#215;</a>
+  <a class="close-reveal-modal" aria-label="Close" role="button" tabindex="0">&#215;</a>
```
Comment (gerardkcohen) [inline doc/includes/reveal/examples_reveal_modal_examples.html]: This should be a tag. aria-haspopup and role="button" not needed. tabindex="0" not needed because anchor tags are already in the tab order.

File: doc/pages/components/reveal.html [hunk 9]
```diff
@@ -124,7 +124,8 @@ <h3 class="subheader">Modal dialogs, or pop-up windows, are handy for prototypin
 
 ## Accessibility
 
-<p class="panel">This component is not yet fully accessible. While it is usable via the keyboard, it has to be checked if additional ARIA attributes can enhance the components accessibility.</p>
+<p class="panel">This component is not yet fully accessible. While it is usable via the keyboard, it has to be checked if additional (ARIA) attributes can enhance the components accessibility. The examples above show how to implement accessible modals using some of those attributes. 
+<br />Please note that you should either add the tabindex- or the href-attribute to anchors to make them fully utilizable by keyboards.</p>
```
Comment (gerardkcohen) [inline doc/pages/components/reveal.html]: Adding tabindex and href-attribute is not enough for an anchor to become accessible as button. I would remove this line.

File: js/foundation/foundation.reveal.js [hunk 10]
```diff
@@ -68,7 +68,7 @@
         });
 
       S(document)
-        .on('click.fndtn.reveal', this.close_targets(), function (e) {
+        .on('click.fndtn.reveal keydown.fndtn.reveal', this.close_targets(), function (e) {
```
Comment (gerardkcohen) [inline js/foundation/foundation.reveal.js]: Should really only be `ENTER` or `SPACE` as not all keys are expected to activate a button

File: doc/includes/reveal/examples_reveal_basic_rendered.html (hunk 4 above)
Comment (Owlbertz) [inline doc/includes/reveal/examples_reveal_basic_rendered.html]: This is not really true for anchors without href. They are skipped otherwise at least in some browsers.

File: doc/pages/components/reveal.html (hunk 9 above)
Comment (citelao) [inline doc/pages/components/reveal.html]: I would suggest writing out more, rather than less, about how to make links and buttons and modals accessible, and make this a fully fledged section in the page, outside of the `.panel`. 

Basically:

- `role="dialog"`
- `<button>` (or `role="button"`)
- `aria-label` or `aria-labelledby`
- `aria-describedby`
- `aria-hidden`

**Comment by citelao** [comment]
:+1:

We should probably include an accessibility writeup in the doc and a focus-eater extension to the js.

As @gerardkcohen said, changing the `<a>` to a `<button>` is important. I am able to take a look at the SASS if need be.

**Comment by zurbrandon** [comment]
Hey guys! This is awesome, we'd love to pull it own and start testing, would you consider this complete?

**Comment by Owlbertz** [comment]
Yes, based on the feedback from #6572 I'd consider this as a valid solution.

**Comment by citelao** [comment]
Yeah, I think this should work. I would write up a bit about how to make an accessible modal and link to [some](http://accessibility.oit.ncsu.edu/blog/2013/09/13/the-incredible-accessible-modal-dialog/) [sites](http://www.smashingmagazine.com/2014/09/15/making-modal-windows-better-for-everyone/), but the code should work.

(I only have Voiceover to test with)

**Comment by Owlbertz** [comment]
Since there seems to not be much more happening in here, I will just close the PR.

**Comment by rafibomb** [comment]
Can you re-open this? We have an accessible modal for Foundation 6 but would be awesome to have it for 5. We have time to test it an merge it now. Thanks!

**Comment by rafibomb** [comment]
Reference https://github.com/zurb/foundation/issues/6579

**Comment by Owlbertz** [comment]
Hi,

**Comment by Owlbertz** [comment]
Hi, nothing happend here for a while so I closed this PR.
Unfortunately it seems that I closed it again by accident just now...
However, I reopen it so you can use it for Foundation 5.
If there is anything to do for Foundation 6 regarding accessibility I'd be happy to help as well.

**Comment by citelao** [comment]
:+1: what's the status of this?

**Comment by rafibomb** [comment]
Thanks for the PR and comments!

@Owlbertz shoot me an email so we can get you on the Foundation 6 private release!
